rocketpy/control/controller.py

```diff
@@ -0,0 +1,132 @@
+               the controller function can interact with. The objects are
+               listed in the same order as they are provided in the
+               `interactable_objects`.
+
+            This function will be called during the simulation at the specified
+            sampling rate. The function should evaluate and change the interactable
+            objects as needed. The function return statement can be used to save
+            relevant information in the `observed_variables` list.
+
+            .. note:: The function will be called according to the sampling rate
+            specified.
+        sampling_rate : float
+            The sampling rate of the controller function in Hertz (Hz). This
+            means that the controller function will be called every
+            `1/sampling_rate` seconds.
+        initial_observed_variables : list, optional
+            A list of the initial values of the variables that the controller
+            function returns. This list is used to initialize the
+            `observed_variables` argument of the controller function. The
+            default value is None, which initializes the list as an empty list.
+        name : str
+            The name of the controller. This will be used for printing and
+            plotting.
+
+        Returns
+        -------
+        None
+        """
+        self.interactable_objects = interactable_objects
+        self.controller_function = controller_function
+        self.sampling_rate = sampling_rate
+        self.name = name
+        self.prints = _ControllerPrints(self)
+
+        if initial_observed_variables is not None:
+            self.observed_variables = [initial_observed_variables]
+        else:
+            self.observed_variables = []
+
+    def __call__(self, time, state_vector, state_history):
+        """Call the controller function. This is used by the simulation class.
+
+        Parameters
+        ----------
+        time : float
+            The time of the simulation in seconds.
+        state_vector : list
+            The state vector of the simulation, which is defined as:
+
+            `[x, y, z, vx, vy, vz, e0, e1, e2, e3, wx, wy, wz]`.
+        state_history : list
+            A list containing the state history of the simulation. The state
+            history is a list of every state vector of every step of the
+            simulation. The state history is a list of lists, where each
+            sublist is a state vector and is ordered from oldest to newest.
+
+        Returns
+        -------
+        None
+        """
+        observed_variables = self.controller_function(
+            time,
+            self.sampling_rate,
+            state_vector,
+            state_history,
+            self.observed_variables,
+            self.interactable_objects,
+        )
+        if observed_variables is not None:
+            self.observed_variables.append(observed_variables)
+
```

Comment:
Just to double check here...
It is really the intention to not include a "else" statement?

Something like:

else:
            self.observed_variables = []

Reply:
Yes, that is the intended behavior